default/scripting/techs/ship_hulls/SHP_GAL_EXPLO.focs.py

```diff
@@ -55,14 +54,22 @@
                 LocalCandidate.Population
                 + LocalCandidate.Population
                 * GROWTH_RATE_FACTOR
-                * (LocalCandidate.TargetPopulation + 1 - LocalCandidate.Population)
+                * (1 + (1 - LocalCandidate.TargetPopulation) / LocalCandidate.TargetPopulation)
```

Comment:
Typically, `SomePlanet.TargetPopulation` will be a number like `20` so. If you subtract that from `1`, you get a negative number, say `-19`. If you divide that by the the original number, you'd get something like `-0.95`. If you add that to `1`, you get a number like `0.05` which is almost always going to be less than `MIN_RECOLONIZING_SIZE`.

Reply:
It should be
`* (1 + (1 - LocalCandidate.Population) / LocalCandidate.TargetPopulation)`
i.e. Population instead of TargetPopulation in the first LocalCandidate.

After this change I get the following error in command line:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  ERROR: Tech "SHP_IMPROVED_ENGINE_COUPLINGS" requires a missing or malformed tech "SHP_GAL_EXPLO" as its prerequisite.
Aborted (core dumped)
```
In the client's log I can find this:

```
13:08:22.603079 {0x00007f38313ab700} [error] client : PythonCommon.cpp:101 : Traceback (most recent call last):
13:08:22.603093 {0x00007f38313ab700} [error] client : PythonCommon.cpp:101 :   File "<string>", line 71, in <module>
13:08:22.603098 {0x00007f38313ab700} [error] client : PythonCommon.cpp:101 : TypeError: '<' not supported between instances of 'float' and 'ValueRefDouble'
13:08:22.603108 {0x00007f38313ab700} [debug] timer : ScopedTimer.cpp:48 : parse_file "SHP_GAL_EXPLO.focs.py" time:     3853 µs
```

The only instance of float vs ValueRef in my code that uses `<`, which was present without this PR, is the following:
```
                ((0.1 < LocalCandidate.Happiness) & (LocalCandidate.Happiness < (MIN_RECOLONIZING_HAPPINESS - 0.001)))
                | ((0.1 < LocalCandidate.Population) & (LocalCandidate.Population < (MIN_RECOLONIZING_SIZE - 0.001)))
```

0.1 is a float, LocalCandidate.Happiness is a ValueRedDouble.

I have no idea how to fix this error. Python doesn't even have double type (Python's float is good enough, otherwise numpy).

@o01eg do you have an insight on what's going on here?

Reply:
@Oberlus There is PR: https://github.com/freeorion/freeorion/pull/4153